build-tools/packages/build-cli/src/commands/merge/branch.ts

```diff
@@ -0,0 +1,214 @@
+        this.log(
+            `${lastMergedCommit} is the last merged commit id between ${flags.source} and ${flags.target}`,
+        );
+
+        const listCommits = await gitRepo.revList(lastMergedCommit, flags.source);
+        const unmergedCommitList: string[] = [];
+        let str = "";
+
+        for (const id of listCommits) {
+            // check the length to be 40 since git commit id is 40 digits long
+            if (str.length === 40) {
+                unmergedCommitList.push(str);
+                str = "";
+                continue;
+            }
+
+            str += id;
+        }
+
+        this.log(
+            `There are ${unmergedCommitList.length} unmerged commits between ${flags.source} and ${flags.target} branches`,
+        );
+
+        if (unmergedCommitList.length === 0) {
+            this.log(
+                `${flags.source} and ${flags.target} branches are in sync. No commits to merge`,
+            );
+            this.exit(-1);
+        }
+
+        const commitSize =
+            flags.batchSize <= unmergedCommitList.length
+                ? flags.batchSize
+                : unmergedCommitList.length;
+        const commitToReset = unmergedCommitList[commitSize - 1];
+        const branchName = `${flags.source}-${flags.target}-${commitToReset}`;
+
+        await gitRepo.switchBranch(flags.target);
+        await gitRepo.createBranch(branchName);
+        await gitRepo.fetchBranch(flags.target, branchName);
+        await gitRepo.setUpstream(branchName);
+
+        // passed a set of commit ids, return the commit id upto which the branch should get reset to
+        let i = 0;
+        let commit = 0;
+        while(i !== commitSize) {
+            try {
+                // eslint-disable-next-line no-await-in-loop
+                await gitRepo.merge(unmergedCommitList[i]);
+            } catch (error: unknown) {
+                // eslint-disable-next-line no-await-in-loop
+                await gitRepo.mergeAbort();
+                this.log(`Merge abort for ${unmergedCommitList[i]}: `, error);
+                break;
+            }
+
+            commit = i;
+            i++;
+        }
+
+        let prNumber;
+        if(i === 0) {
+            this.log(`Opening a pull request for a single commit id ${unmergedCommitList[0]} as it might have merge conflicts with the ${flags.target} branch.`);
+            await gitRepo.switchBranch(flags.source);
+            await gitRepo.createBranch(`${flags.source}-${flags.target}-${unmergedCommitList[0]}`);
+            await gitRepo.fetchBranch(flags.source, `${flags.source}-${flags.target}-${unmergedCommitList[0]}`);
+            await gitRepo.setUpstream(`${flags.source}-${flags.target}-${unmergedCommitList[0]}`);
+            await gitRepo.resetBranch(unmergedCommitList[0]);
+            prNumber = await createPR(flags.auth, `${flags.source}-${flags.target}-${unmergedCommitList[0]}`, flags.target, "sonalivdeshpande");
```

Comment:
should your username be here?

Reply:
Tested it locally and username is not required for any of the git apis. The only problem I'm seeing is `git rev-list` is not returning the right commit ids on my fork.